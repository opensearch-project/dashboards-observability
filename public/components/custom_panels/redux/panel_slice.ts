/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { createSelector, createSlice } from '@reduxjs/toolkit';
import { concat, from, Observable, of } from 'rxjs';
import { map, mergeMap, toArray } from 'rxjs/operators';
import { forEach } from 'lodash';
import {
  createDemoPanel,
  CUSTOM_PANELS_API_PREFIX,
  CUSTOM_PANELS_SAVED_OBJECT_TYPE,
  samplePanelName,
} from '../../../../common/constants/custom_panels';
import {
  CustomPanelListType,
  CustomPanelType,
  ObservabilityPanelAttrs,
  PanelType,
} from '../../../../common/types/custom_panels';
import { coreRefs } from '../../../framework/core_refs';
import { SavedObject, SimpleSavedObject } from '../../../../../../src/core/public';
import { addVisualizationPanel } from '../helpers/add_visualization_helper';

interface InitialState {
  id: string;
  panel: CustomPanelType;
  panelList: CustomPanelType[];
}

export const newPanelTemplate = (newName): PanelType => ({
  title: newName,
  dateCreated: new Date().getTime(),
  dateModified: new Date().getTime(),
  visualizations: [],
  queryFilter: { language: '', query: '' },
  timeRange: { from: 'now-1d', to: 'now' },
});

const initialState: InitialState = {
  id: '',
  panel: newPanelTemplate(''),
  panelList: [],
  loadingFlag: false,
};

export const panelSlice = createSlice({
  name: 'customPanel',
  initialState,
  reducers: {
    setPanelId: (state, action) => ({ ...state, id: action.payload }),

    setPanel: (state, action) => {
      return { ...state, panel: action.payload };
    },

    setPanelList: (state, action) => {
      return { ...state, panelList: action.payload };
    },
  },
});

export const { setPanel, setPanelList } = panelSlice.actions;

export const panelReducer = panelSlice.reducer;

export const selectPanel = createSelector(
  (rootState) => rootState.customPanel.panel,
  (panel) => normalizedPanel(panel)
);

const normalizedPanel = (panel: CustomPanelType): CustomPanelType => ({
  ...newPanelTemplate(''),
  ...panel,
});

export const selectPanelList = (rootState): CustomPanelType[] => rootState.customPanel.panelList;

/*
 ** ASYNC DISPATCH FUNCTIONS
 */

const fetchSavedObjectPanels$ = () =>
  from(savedObjectPanelsClient.find()).pipe(
    mergeMap((res) => res.savedObjects),
    map(savedObjectToCustomPanel)
  );

const fetchObservabilityPanels$ = () =>
  of(coreRefs.http.get(`${CUSTOM_PANELS_API_PREFIX}/panels`)).pipe(
    mergeMap((res) => res),
    mergeMap((res) => res.panels as ObservabilityPanelAttrs[]),
    map((p: ObservabilityPanelAttrs) => ({
      ...p,
      title: p.name,
      savedObject: false,
      type: 'observability-savedObject',
    }))
  );

// Fetches all saved Custom Panels
const fetchCustomPanels = async () => {
  const panels$: Observable<CustomPanelListType> = concat(
    fetchSavedObjectPanels$(),
    fetchObservabilityPanels$()
  ).pipe(
    map((res) => {
      return res as CustomPanelListType;
    })
  );

  return panels$.pipe(toArray()).toPromise();
};

export const fetchPanels = () => async (dispatch, getState) => {
  const panels = await fetchCustomPanels();
  dispatch(setPanelList(panels));
};

export const fetchPanel = (id) => async (dispatch, getState) => {
  const soPanel = await savedObjectPanelsClient.get(id);
  const panel = savedObjectToCustomPanel(soPanel);
  dispatch(setPanel(panel));
};

export const fetchVisualization = () => (dispatch, getState) => {};

const updateLegacyPanel = (panel: CustomPanelType) =>
  coreRefs.http!.post(`${CUSTOM_PANELS_API_PREFIX}/panels/update`, {
    body: JSON.stringify({ panelId: panel.id, panel: panel as PanelType }),
  });

const updateSavedObjectPanel = (panel: CustomPanelType) => savedObjectPanelsClient.update(panel);

export const uuidRx = /^[0-9a-fA-F]{8}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{12}$/;

export const isUuid = (id) => !!id.match(uuidRx);

export const updatePanel = (
  panel: CustomPanelType,
  successMsg: string,
  failureMsg: string
) => async (dispatch, getState) => {
  const { toasts } = coreRefs;

  try {
    if (isUuid(panel.id)) await updateSavedObjectPanel(panel);
    else await updateLegacyPanel(panel);
    if (successMsg) {
      toasts!.add(successMsg);
    }
    dispatch(setPanel(panel));
    const panelList = getState().customPanel.panelList.map((p) => (p.id === panel.id ? panel : p));
    dispatch(setPanelList(panelList));
  } catch (e) {
    if (failureMsg) {
      toasts!.addDanger(failureMsg);
    }
    console.error(e);
  }
};

export const addVizToPanels = (panels, vizId) => async (dispatch, getState) => {
  forEach(panels, (oldPanel) => {
    const panel = getState().customPanel.panelList.find((p) => p.id === oldPanel.panel.id);

    const allVisualizations = panel!.visualizations;

    const visualizationsWithNewPanel = addVisualizationPanel(vizId, undefined, allVisualizations);

    const updatedPanel = { ...panel, visualizations: visualizationsWithNewPanel };
    dispatch(updatePanel(updatedPanel, '', ''));
  });
};

export const addMultipleVizToPanels = async (panels, vizIds) => {
  await Promise.all(
    panels.map(async ({ panel: { id } }) => {
      const soPanel = await savedObjectPanelsClient.get(id);
      const oldPanel = savedObjectToCustomPanel(soPanel);

      const allVisualizations = oldPanel!.visualizations;

      let visualizationsWithNewPanel = allVisualizations;

      forEach(vizIds, (vizId) => {
        visualizationsWithNewPanel = addVisualizationPanel(
          vizId,
          undefined,
          visualizationsWithNewPanel
        );
      });

      const updatedPanel = { ...oldPanel, visualizations: visualizationsWithNewPanel };

      if (isUuid(updatedPanel.id)) {
        await updateSavedObjectPanel(updatedPanel);
      } else {
        await updateLegacyPanel(updatedPanel);
      }
    })
  );
};

export const replaceVizInPanel = (oldPanel, oldVizId, vizId, newVisualizationTitle) => async (
  dispatch,
  getState
) => {
  const panel = getState().customPanel.panelList.find((p) => p.id === oldPanel.id);

  const allVisualizations = panel!.visualizations;

  const visualizationsWithNewPanel = addVisualizationPanel(vizId, oldVizId, allVisualizations);

  const updatedPanel = { ...panel, visualizations: visualizationsWithNewPanel };

  dispatch(
    updatePanel(
      updatedPanel,
      `Visualization ${newVisualizationTitle} successfully added!`,
      `Error in adding ${newVisualizationTitle} visualization to the panel`
    )
  );
};

const deletePanelSO = (customPanelIdList: string[]) => {
  const soPanelIds = customPanelIdList.filter((id) => isUuid(id));
  return Promise.all(soPanelIds.map((id) => savedObjectPanelsClient.delete(id)));
};

const deleteLegacyPanels = (customPanelIdList: string[]) => {
  const panelIds = customPanelIdList.filter((id) => !isUuid(id));
  if (panelIds.length === 0) return;

  const concatList = panelIds.toString();
  return coreRefs.http!.delete(`${CUSTOM_PANELS_API_PREFIX}/panelList/` + concatList);
};

export const deletePanels = (panelsToDelete: CustomPanelType[]) => async (dispatch, getState) => {
  const { toasts } = coreRefs;

  const toastMessage = `Observability Dashboard${
    panelsToDelete.length > 1 ? 's' : ' ' + panelsToDelete[0].title
  } successfully deleted!`;
  try {
    const ids = panelsToDelete.map((p) => p.id);
    await Promise.all([deleteLegacyPanels(ids), deletePanelSO(ids)]);

    const panelList: CustomPanelType[] = getState().customPanel.panelList.filter(
      (p) => !ids.includes(p.id)
    );
    dispatch(setPanelList(panelList));
    toasts!.add(toastMessage);
  } catch (e) {
    toasts!.addDanger(
      'Error deleting Observability Dashboards, please make sure you have the correct permission.'
    );
    console.error(e);
  }
};

export const createPanel = (panel) => async (dispatch, getState) => {
  const { toasts } = coreRefs;
  try {
    const newSOPanel = await savedObjectPanelsClient.create(panel);
    const newPanel = savedObjectToCustomPanel(newSOPanel);
    const panelList = getState().customPanel.panelList;
    dispatch(setPanelList([...panelList, newPanel]));
    toasts!.add(`Observability Dashboard "${newPanel.title}" successfully created!`);
    window.location.replace(`#/${newPanel.id}`);
  } catch (e) {
    toasts!.addDanger(
      'Error occurred while creating Observability Dashboard, please make sure you have the correct permission.'
    );
    console.error(e);
  }
};

export const createPanelSample = (vizIds) => async (dispatch, getState) => {
  const samplePanel = {
    ...createDemoPanel(vizIds),
    dateCreated: new Date().getTime(),
    dateModified: new Date().getTime(),
    title: samplePanelName,
  };
  const newSOPanel = await savedObjectPanelsClient.create(samplePanel);
  const newPanel = savedObjectToCustomPanel(newSOPanel);
  const panelList = getState().customPanel.panelList;
  dispatch(setPanelList([...panelList, newPanel]));
};

export const clonePanel = (panel, newPanelName) => async (dispatch, getState) => {
  const { toasts } = coreRefs;

  try {
    const { id, ...panelCopy } = {
      ...panel,
      title: newPanelName,
      dateCreated: new Date().getTime(),
      dateModified: new Date().getTime(),
    } as PanelType;

    const newSOPanel = await savedObjectPanelsClient.create(panelCopy);

    const newPanel = savedObjectToCustomPanel(newSOPanel);
    const panelList = getState().customPanel.panelList;
    dispatch(setPanelList([...panelList, newPanel]));
    dispatch(setPanel(newPanel));
    toasts!.add(`Observability Dashboard "${newPanel.title}" successfully created!`);
    window.location.replace(`#/${newPanel.id}`);
  } catch (e) {
    toasts!.addDanger(
      'Error cloning Observability Dashboard, please make sure you have the correct permission.'
    );
    console.error(e);
  }
};

const saveRenamedPanel = async (id, name) => {
  const renamePanelObject = {
    panelId: id,
    panelName: name,
  };

  return http.post(`${CUSTOM_PANELS_API_PREFIX}/panels/rename`, {
    body: JSON.stringify(renamePanelObject),
  });
};

const saveRenamedPanelSO = async (id, name) => {
  const panel: SavedObject<PanelType> = await coreRefs.savedObjectsClient!.get(
    CUSTOM_PANELS_SAVED_OBJECT_TYPE,
    id
  );
  panel.title = name;
  await coreRefs.savedObjectsClient!.update(CUSTOM_PANELS_SAVED_OBJECT_TYPE, id, panel);
};

// Renames an existing CustomPanel
export const renameCustomPanel = (editedCustomPanelName: string, id: string) => async (
  dispatch,
  getState
) => {
  const panel = getState().customPanel.panelList.find((p) => p.id === id);
  const updatedPanel = { ...panel, title: editedCustomPanelName };
  dispatch(
    updatePanel(
      updatedPanel,
      `Operational Panel successfully renamed into "${editedCustomPanelName}"`,
      'Error renaming Operational Panel, please make sure you have the correct permission.'
    )
  );
};

/*
 ** UTILITY FUNCTIONS
 */
const savedObjectToCustomPanel = (so: SimpleSavedObject<PanelType>): CustomPanelType => ({
  id: so.id,
  type: so.type,
  objectId: so.type + ':' + so.id,
  ...so.attributes,
  savedObject: true,
});

const savedObjectPanelsClient = {
  find: (options) =>
    coreRefs.savedObjectsClient!.find({ type: CUSTOM_PANELS_SAVED_OBJECT_TYPE, ...options }),
  delete: (id) => coreRefs.savedObjectsClient!.delete(CUSTOM_PANELS_SAVED_OBJECT_TYPE, id),
  update: (panel) =>
    coreRefs.savedObjectsClient!.update(CUSTOM_PANELS_SAVED_OBJECT_TYPE, panel.id, panel),
  get: (id) => coreRefs.savedObjectsClient!.get(CUSTOM_PANELS_SAVED_OBJECT_TYPE, id),
  create: (panel) => coreRefs.savedObjectsClient!.create(CUSTOM_PANELS_SAVED_OBJECT_TYPE, panel),
};
