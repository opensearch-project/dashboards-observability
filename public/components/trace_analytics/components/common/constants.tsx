/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

// Conversion factor for nanoseconds to milliseconds
export const NANOS_TO_MS = 1e6;

export const MILI_TO_SEC = 1000;

export const pieChartColors = [
  '#7492e7',
  '#c33d69',
  '#2ea597',
  '#8456ce',
  '#e07941',
  '#3759ce',
  '#ce567c',
  '#9469d6',
  '#4066df',
  '#da7596',
];

export interface Span {
  traceId: string;
  spanId: string;
  traceState: string;
  parentSpanId: string;
  name: string;
  kind: string;
  startTime: string;
  endTime: string;
  durationInNanos: number;
  serviceName: string;
  events: any[];
  links: any[];
  droppedAttributesCount: number;
  droppedEventsCount: number;
  droppedLinksCount: number;
  traceGroup: string;
  traceGroupFields: {
    endTime: string;
    durationInNanos: number;
    statusCode: number;
  };
  status: {
    code: number;
  };
  instrumentationLibrary: {
    name: string;
    version: string;
  };
}

export interface ParsedHit {
  _index: string;
  _id: string;
  _score: number;
  _source: Span;
  sort?: any[];
}
