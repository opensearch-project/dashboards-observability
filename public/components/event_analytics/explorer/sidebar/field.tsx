/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  EuiBadge,
  EuiButtonIcon,
  EuiFlexGroup,
  EuiFlexItem,
  EuiLoadingSpinner,
  EuiPopover,
  EuiText,
  EuiTitle,
  EuiToolTip,
} from '@elastic/eui';
import { i18n } from '@osd/i18n';
import { isEqual, toUpper, upperFirst } from 'lodash';
import React, { useEffect, useState } from 'react';
import { useSelector } from 'react-redux';
import { DEFAULT_DATA_SOURCE_TYPE } from '../../../../../common/constants/data_sources';
import { IField } from '../../../../../common/types/explorer';
import { FieldIcon } from '../../../common/field_icon';
import { selectSearchMetaData } from '../../../event_analytics/redux/slices/search_meta_data_slice';
import { FieldInsights } from './field_insights';

interface IFieldProps {
  query: string;
  field: IField;
  selectedPattern: string;
  isOverridingPattern: boolean;
  handleOverridePattern: (pattern: IField) => void;
  selectedTimestamp: string;
  isOverridingTimestamp: boolean;
  handleOverrideTimestamp: (timestamp: IField) => void;
  selected: boolean;
  showToggleButton: boolean;
  showTimestampOverrideButton: boolean;
  isFieldToggleButtonDisabled: boolean;
  onToggleField: (field: IField) => void;
  tabId: string;
}

export const Field = (props: IFieldProps) => {
  const {
    query,
    field,
    selectedPattern,
    isOverridingPattern,
    handleOverridePattern,
    selectedTimestamp,
    isOverridingTimestamp,
    handleOverrideTimestamp,
    selected,
    isFieldToggleButtonDisabled = false,
    showTimestampOverrideButton = true,
    onToggleField,
    tabId,
  } = props;

  const [isFieldDetailsOpen, setIsFieldDetailsOpen] = useState(false);
  const explorerSearchMeta = useSelector(selectSearchMetaData)[tabId] || {};
  const isDefaultDataSourceType =
    explorerSearchMeta.datasources?.[0]?.type === DEFAULT_DATA_SOURCE_TYPE;
  const appLogEvents = tabId.startsWith('application-analytics-tab');

  const addLabelAria = i18n.translate('addButtonAriaLabel', {
    defaultMessage: 'Add {field} to table',
    values: { field: field.name },
  });
  const removeLabelAria = i18n.translate('removeButtonAriaLabel', {
    defaultMessage: 'Remove {field} from table',
    values: { field: field.name },
  });

  const togglePopover = () => {
    setIsFieldDetailsOpen((staleState) => !staleState);
  };

  const toggleField = (fields: IField) => {
    onToggleField(fields);
  };

  // hardcoded for demo purposes - remove afterwards
  useEffect(() => {
    if (field.name === 'order_date') {
      handleOverrideTimestamp(field);
    }
  }, []);

  return (
    <EuiFlexGroup gutterSize="s" alignItems="center" responsive={false} className="dscSidebarField">
      <EuiFlexItem grow={false}>
        <FieldIcon type={isEqual(field.type, 'timestamp') ? 'date' : field.type} />
      </EuiFlexItem>
      <EuiFlexItem grow>
        <EuiText size="xs">{field.name}</EuiText>
      </EuiFlexItem>
      <EuiToolTip id="override-pattern" delay="long" content="Override default pattern">
        <>
          {isEqual(field.type, 'string') ? (
            isEqual(selectedPattern, field.name) ? (
              <EuiText size="xs">
                {' '}
                <EuiFlexItem grow={false}>
                  <EuiBadge
                    className="dscSidebarField__actionButton"
                    color="hollow"
                    data-test-subj="eventFields__default-pattern-mark"
                  >
                    Default Pattern
                  </EuiBadge>
                </EuiFlexItem>
              </EuiText>
            ) : isOverridingPattern ? (
              <EuiFlexItem grow={false}>
                <EuiLoadingSpinner className="override_pattern_loading" size="s" />
              </EuiFlexItem>
            ) : (
              <EuiFlexItem grow={false}>
                <EuiButtonIcon
                  aria-labelledby="override_pattern"
                  size="xs"
                  color="text"
                  iconType="inputOutput"
                  onClick={() => handleOverridePattern(field)}
                  data-test-subj="eventExplorer__overrideDefaultPattern"
                  className="dscSidebarField__actionButton"
                  isDisabled={!(isDefaultDataSourceType || appLogEvents)}
                >
                  Override
                </EuiButtonIcon>
              </EuiFlexItem>
            )
          ) : null}
        </>
      </EuiToolTip>
      <EuiToolTip id="override-timestamp" delay="long" content="Override default timestamp">
        <>
          {showTimestampOverrideButton && isEqual(field.type, 'timestamp') ? (
            isEqual(selectedTimestamp, field.name) ? (
              <EuiFlexItem grow={false}>
                <EuiText size="xs">
                  {' '}
                  <EuiBadge
                    className="dscSidebarField__actionButton"
                    color="hollow"
                    data-test-subj="eventFields__default-timestamp-mark"
                  >
                    Default Timestamp
                  </EuiBadge>
                </EuiText>
              </EuiFlexItem>
            ) : isOverridingTimestamp ? (
              <EuiFlexItem grow={false}>
                <EuiLoadingSpinner className="override_timestamp_loading" size="s" />
              </EuiFlexItem>
            ) : (
              <EuiFlexItem grow={false}>
                <EuiButtonIcon
                  aria-labelledby="override_timestamp"
                  size="xs"
                  color="text"
                  iconType="inputOutput"
                  onClick={() => handleOverrideTimestamp(field)}
                  data-test-subj="eventExplorer__overrideDefaultTimestamp"
                  className="dscSidebarField__actionButton"
                  isDisabled={!(isDefaultDataSourceType || appLogEvents)}
                >
                  Override
                </EuiButtonIcon>
              </EuiFlexItem>
            )
          ) : null}
        </>
      </EuiToolTip>
      <EuiFlexItem grow={false}>
        <EuiToolTip delay="long" content="inspect">
          <EuiPopover
            ownFocus
            display="block"
            isOpen={isFieldDetailsOpen}
            closePopover={() => setIsFieldDetailsOpen(false)}
            anchorPosition="rightUp"
            panelClassName="explorerSidebarItem__fieldPopoverPanel"
            button={
              <EuiButtonIcon
                iconType="inspect"
                size="xs"
                onClick={togglePopover}
                aria-label={'inspect'}
                className="dscSidebarField__actionButton"
                isDisabled={!(isDefaultDataSourceType || appLogEvents)}
              />
            }
          >
            <EuiFlexGroup justifyContent="spaceBetween">
              <EuiFlexItem>
                <EuiTitle size="xs">
                  <h4>{toUpper(field.name)}</h4>
                </EuiTitle>
              </EuiFlexItem>
              <EuiFlexItem grow={false}>{upperFirst(field.type)}</EuiFlexItem>
            </EuiFlexGroup>

            <FieldInsights field={field} query={query} />
          </EuiPopover>
        </EuiToolTip>
      </EuiFlexItem>
      <EuiFlexItem grow={false}>
        <EuiToolTip
          delay="long"
          content={
            isFieldToggleButtonDisabled
              ? "Toggle button is disabled on query contains 'stats' or no hits for the search"
              : selected
              ? 'Remove field from table'
              : 'Add field as column'
          }
        >
          <>
            {isFieldToggleButtonDisabled ? (
              <EuiButtonIcon
                color={selected ? 'danger' : 'primary'}
                iconType={selected ? 'cross' : 'plusInCircleFilled'}
                isDisabled
                data-test-subj={`fieldToggle-${field.name}`}
                aria-label={selected ? removeLabelAria : addLabelAria}
                className="dscSidebarField__actionButton"
              />
            ) : (
              <EuiButtonIcon
                color={selected ? 'danger' : 'primary'}
                iconType={selected ? 'cross' : 'plusInCircleFilled'}
                onClick={(e: React.MouseEvent<HTMLButtonElement>) => {
                  if (e.type === 'click') {
                    e.currentTarget.focus();
                  }
                  e.preventDefault();
                  e.stopPropagation();
                  toggleField(field);
                }}
                data-test-subj={`fieldToggle-${field.name}`}
                aria-label={selected ? removeLabelAria : addLabelAria}
                className="dscSidebarField__actionButton"
              />
            )}
          </>
        </EuiToolTip>
      </EuiFlexItem>
    </EuiFlexGroup>
  );
};
