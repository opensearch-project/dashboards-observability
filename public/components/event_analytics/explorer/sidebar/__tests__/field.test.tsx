/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import { waitFor } from '@testing-library/react';
import { Field } from '../field';
import { AGENT_FIELD } from '../../../../../../test/event_analytics_constants';
import { applyMiddleware, createStore } from 'redux';
import { rootReducer } from '../../../../../framework/redux/reducers';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import { DEFAULT_DATA_SOURCE_TYPE } from '../../../../../../common/constants/data_sources';

describe('Field component', () => {
  configure({ adapter: new Adapter() });

  it('Renders a sidebar field', async () => {
    const onToggleField = jest.fn();
    const handleOverrideTimestamp = jest.fn();
    const selectedTimestamp = 'timestamp';
    const store = createStore(rootReducer, applyMiddleware(thunk));

    const wrapper = mount(
      <Provider store={store}>
        <Field
          field={AGENT_FIELD}
          selectedTimestamp={selectedTimestamp}
          handleOverrideTimestamp={handleOverrideTimestamp}
          isOverridingTimestamp={false}
          isFieldToggleButtonDisabled={false}
          showTimestampOverrideButton={true}
          onToggleField={onToggleField}
          selected
          showToggleButton={true}
          tabId={DEFAULT_DATA_SOURCE_TYPE}
        />
      </Provider>
    );

    wrapper.update();

    await waitFor(() => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
