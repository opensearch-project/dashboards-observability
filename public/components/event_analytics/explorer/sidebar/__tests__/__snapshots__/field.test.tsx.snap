// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field component Renders a sidebar field 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(Symbol.observable): [Function],
    }
  }
>
  <Component
    field={
      Object {
        "name": "agent",
        "type": "string",
      }
    }
    handleOverrideTimestamp={[MockFunction]}
    isFieldToggleButtonDisabled={false}
    isOverridingTimestamp={false}
    onToggleField={[MockFunction]}
    selected={true}
    selectedTimestamp="timestamp"
    showTimestampOverrideButton={true}
    showToggleButton={true}
    tabId="DEFAULT_INDEX_PATTERNS"
  >
    <EuiFlexGroup
      alignItems="center"
      className="dscSidebarField"
      gutterSize="s"
      responsive={false}
    >
      <div
        className="euiFlexGroup euiFlexGroup--gutterSmall euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow dscSidebarField"
      >
        <EuiFlexItem
          grow={false}
        >
          <div
            className="euiFlexItem euiFlexItem--flexGrowZero"
          >
            <FieldIcon
              type="string"
            >
              <EuiToken
                aria-label="string"
                className="osdFieldIcon"
                iconType="tokenString"
                size="s"
                title="string"
              >
                <span
                  className="euiToken euiToken--euiColorVis1 euiToken--square euiToken--light euiToken--small osdFieldIcon"
                  style={Object {}}
                >
                  <EuiIcon
                    aria-label="string"
                    size="m"
                    title="string"
                    type="tokenString"
                  >
                    <EuiIconEmpty
                      aria-hidden={true}
                      aria-label="string"
                      className="euiIcon euiIcon--medium euiIcon-isLoading"
                      focusable="false"
                      role="img"
                      style={null}
                      title="string"
                    >
                      <svg
                        aria-hidden={true}
                        aria-label="string"
                        className="euiIcon euiIcon--medium euiIcon-isLoading"
                        focusable="false"
                        height={16}
                        role="img"
                        style={null}
                        viewBox="0 0 16 16"
                        width={16}
                        xmlns="http://www.w3.org/2000/svg"
                      />
                    </EuiIconEmpty>
                  </EuiIcon>
                </span>
              </EuiToken>
            </FieldIcon>
          </div>
        </EuiFlexItem>
        <EuiFlexItem
          grow={true}
        >
          <div
            className="euiFlexItem"
          >
            <EuiText
              size="xs"
            >
              <div
                className="euiText euiText--extraSmall"
              >
                agent
              </div>
            </EuiText>
          </div>
        </EuiFlexItem>
        <EuiToolTip
          content="Override default pattern"
          delay="long"
          id="override-pattern"
          position="top"
        >
          <span
            className="euiToolTipAnchor"
            onKeyUp={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
          >
            <EuiFlexItem
              grow={false}
            >
              <div
                className="euiFlexItem euiFlexItem--flexGrowZero"
              >
                <EuiButtonIcon
                  aria-labelledby="override_pattern"
                  className="dscSidebarField__actionButton"
                  color="text"
                  data-test-subj="eventExplorer__overrideDefaultPattern"
                  iconType="inputOutput"
                  isDisabled={true}
                  onClick={[Function]}
                  size="xs"
                >
                  <button
                    aria-labelledby="override_pattern"
                    className="euiButtonIcon euiButtonIcon-isDisabled euiButtonIcon--text euiButtonIcon--empty euiButtonIcon--xSmall dscSidebarField__actionButton"
                    data-test-subj="eventExplorer__overrideDefaultPattern"
                    disabled={true}
                    onClick={[Function]}
                    type="button"
                  >
                    <EuiIcon
                      aria-hidden="true"
                      className="euiButtonIcon__icon"
                      color="inherit"
                      size="m"
                      type="inputOutput"
                    >
                      <EuiIconEmpty
                        aria-hidden={true}
                        className="euiIcon euiIcon--medium euiIcon--inherit euiIcon-isLoading euiButtonIcon__icon"
                        focusable="false"
                        role="img"
                        style={null}
                      >
                        <svg
                          aria-hidden={true}
                          className="euiIcon euiIcon--medium euiIcon--inherit euiIcon-isLoading euiButtonIcon__icon"
                          focusable="false"
                          height={16}
                          role="img"
                          style={null}
                          viewBox="0 0 16 16"
                          width={16}
                          xmlns="http://www.w3.org/2000/svg"
                        />
                      </EuiIconEmpty>
                    </EuiIcon>
                  </button>
                </EuiButtonIcon>
              </div>
            </EuiFlexItem>
          </span>
        </EuiToolTip>
        <EuiToolTip
          content="Override default timestamp"
          delay="long"
          id="override-timestamp"
          position="top"
        >
          <span
            className="euiToolTipAnchor"
            onKeyUp={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
          />
        </EuiToolTip>
        <EuiFlexItem
          grow={false}
        >
          <div
            className="euiFlexItem euiFlexItem--flexGrowZero"
          >
            <EuiToolTip
              content="inspect"
              delay="long"
              position="top"
            >
              <span
                className="euiToolTipAnchor"
                onKeyUp={[Function]}
                onMouseOut={[Function]}
                onMouseOver={[Function]}
              >
                <EuiPopover
                  anchorPosition="rightUp"
                  button={
                    <EuiButtonIcon
                      aria-label="inspect"
                      className="dscSidebarField__actionButton"
                      iconType="inspect"
                      isDisabled={true}
                      onClick={[Function]}
                      size="xs"
                    />
                  }
                  closePopover={[Function]}
                  display="block"
                  hasArrow={true}
                  isOpen={false}
                  onBlur={[Function]}
                  onFocus={[Function]}
                  ownFocus={true}
                  panelClassName="explorerSidebarItem__fieldPopoverPanel"
                  panelPaddingSize="m"
                >
                  <div
                    className="euiPopover euiPopover--anchorRightUp euiPopover--displayBlock"
                    onBlur={[Function]}
                    onFocus={[Function]}
                  >
                    <div
                      className="euiPopover__anchor"
                    >
                      <EuiButtonIcon
                        aria-label="inspect"
                        className="dscSidebarField__actionButton"
                        iconType="inspect"
                        isDisabled={true}
                        onClick={[Function]}
                        size="xs"
                      >
                        <button
                          aria-label="inspect"
                          className="euiButtonIcon euiButtonIcon-isDisabled euiButtonIcon--primary euiButtonIcon--empty euiButtonIcon--xSmall dscSidebarField__actionButton"
                          disabled={true}
                          onClick={[Function]}
                          type="button"
                        >
                          <EuiIcon
                            aria-hidden="true"
                            className="euiButtonIcon__icon"
                            color="inherit"
                            size="m"
                            type="inspect"
                          >
                            <EuiIconEmpty
                              aria-hidden={true}
                              className="euiIcon euiIcon--medium euiIcon--inherit euiIcon-isLoading euiButtonIcon__icon"
                              focusable="false"
                              role="img"
                              style={null}
                            >
                              <svg
                                aria-hidden={true}
                                className="euiIcon euiIcon--medium euiIcon--inherit euiIcon-isLoading euiButtonIcon__icon"
                                focusable="false"
                                height={16}
                                role="img"
                                style={null}
                                viewBox="0 0 16 16"
                                width={16}
                                xmlns="http://www.w3.org/2000/svg"
                              />
                            </EuiIconEmpty>
                          </EuiIcon>
                        </button>
                      </EuiButtonIcon>
                    </div>
                  </div>
                </EuiPopover>
              </span>
            </EuiToolTip>
          </div>
        </EuiFlexItem>
        <EuiFlexItem
          grow={false}
        >
          <div
            className="euiFlexItem euiFlexItem--flexGrowZero"
          >
            <EuiToolTip
              content="Remove field from table"
              delay="long"
              position="top"
            >
              <span
                className="euiToolTipAnchor"
                onKeyUp={[Function]}
                onMouseOut={[Function]}
                onMouseOver={[Function]}
              >
                <EuiButtonIcon
                  aria-label="Remove agent from table"
                  className="dscSidebarField__actionButton"
                  color="danger"
                  data-test-subj="fieldToggle-agent"
                  iconType="cross"
                  onClick={[Function]}
                >
                  <button
                    aria-label="Remove agent from table"
                    className="euiButtonIcon euiButtonIcon--danger euiButtonIcon--empty euiButtonIcon--xSmall dscSidebarField__actionButton"
                    data-test-subj="fieldToggle-agent"
                    disabled={false}
                    onClick={[Function]}
                    type="button"
                  >
                    <EuiIcon
                      aria-hidden="true"
                      className="euiButtonIcon__icon"
                      color="inherit"
                      size="m"
                      type="cross"
                    >
                      <EuiIconEmpty
                        aria-hidden={true}
                        className="euiIcon euiIcon--medium euiIcon--inherit euiIcon-isLoading euiButtonIcon__icon"
                        focusable="false"
                        role="img"
                        style={null}
                      >
                        <svg
                          aria-hidden={true}
                          className="euiIcon euiIcon--medium euiIcon--inherit euiIcon-isLoading euiButtonIcon__icon"
                          focusable="false"
                          height={16}
                          role="img"
                          style={null}
                          viewBox="0 0 16 16"
                          width={16}
                          xmlns="http://www.w3.org/2000/svg"
                        />
                      </EuiIconEmpty>
                    </EuiIcon>
                  </button>
                </EuiButtonIcon>
              </span>
            </EuiToolTip>
          </div>
        </EuiFlexItem>
      </div>
    </EuiFlexGroup>
  </Component>
</Provider>
`;
